src/encoded/batch_download.py

```diff
@@ -216,13 +247,123 @@ def peak_metadata(context, request):
+    else:
+        try:
+            elements = request.json.get('elements')
+        except ValueError:
+            pass
+        else:
+            param_list['@id'] = elements
+    path = '{}?{}'.format(search_path, urlencode(param_list, True))
+    results = request.embed(path, as_user=True)
+    annotation_graph = results['@graph']
+    ids = [g['@id'] for g in annotation_graph]
+    file_graph = _get_graph('file', request, ids)
+    software_graph = _get_graph('software', request, ids)
+    headers = [header for header in _tsv_mapping_annotation]
+    fout = io.StringIO()
+    writer = csv.writer(fout, delimiter='\t')
+    writer.writerow(headers)
+    for file in file_graph:
+        annotation = _get_embedded_schema_by_dataset(annotation_graph, file['dataset'])
+        software = _get_embedded_schema_by_dataset(software_graph, file['dataset'])
+        writer.writerow([
+            file.get('title', ''),
+            file.get('file_type', ''),
+            file.get('output_type', ''),
+            annotation.get('accession', ''),
+            annotation.get('annotation_type', ''),
+            software.get('title', ''),
+            annotation.get('encyclopedia_version', ''),
+            annotation.get('biosample_term_id', ''),
+            annotation.get('biosample_term_name', ''),
+            annotation.get('biosample_type', ''),
+            annotation.get('relevant_life_stage', ''),
+            annotation.get('relevant_timepoint', ''),
+            annotation.get('relevant_timepoint_units', ''),
+            annotation.get('organism', {}).get('scientific_name', ''),
+            annotation.get('targets', {}).get('name', ''),
+            annotation.get('date_released', ''),
+            annotation.get('award').get('project', ''),
+            file.get('derived_from', ''),
+            file.get('file_size', ''),
+            file.get('lab', {}).get('title', ''),
+            file.get('md5sum', ''),
+            file.get('dbxrefs', ''),
+            file.get('href', ''),
+            file.get('assembly', ''),
+            file.get('controlled_by', ''),
+            file.get('status', ''),
+            file.get('restricted', ''),
+        ])
+    return Response(
+        content_type='text/tsv',
+        body=fout.getvalue(),
+        content_disposition='attachment;filename="%s"' % 'metadata.tsv'
+    )
+
+
+def _get_embedded_schema_by_dataset(graph, dataset):
+    """
+    Get schema based on dataset.
+
+        :param graph: @graph value.
+        :param dataset: dataset.
+    """
+    schema = list(filter(lambda g: g.get('@id') == dataset, graph)) or [{}]
+    return schema[0]
+
+
+def _get_graph(schema_type, request, ids):
+    """
+    Get a graph based on schema type and id.
+
+        :param schema_type: schema type.
+        :param request: request object from pyramid.
+        :param ids: ids.
+    """
+    datasets = ''.join([''.join(['&dataset=', id]) for id in ids])
```

Comment:
how big a search do you create before this is too slow?    Like what about this query: `https://www.encodeproject.org/search/?type=Annotation&encyclopedia_version=4`

Reply:
Do you mean the code, line 347 is slow or do you mean the query size is slow? If it is the code, I can use for-loop rather than list comprehension. I know "join" is faster than using '+'

Reply:
Hmm... I *suspect* it might be slow, was it tested?   How long does it take to create a metadata.tsv for an annotation search with a large number of datasets?   Test it for 10/100/1000/all Annotations.   I think the total number is currently >6,000 and it will probably grow.   Demo is off, but you should check it (it is a good habit).

Reply:
I am still investigating this. There are problems with even 3000 I do not see in experiment.